Speed up GetCorrectedTime and GetSiPMPosition methods by reading the covered runs only once.

cvilelahep opened the pull request:
Speeds up the GetCorrectedTime by about a factor of 6.

olantwin commented on shipLHC/MuFilter.cxx:
```diff
@@ -650,4 +658,15 @@ void MuFilter::GetPosition(Int_t fDetectorID, TVector3& vLeft, TVector3& vRight)
        if (subsystem==1){return conf_ints["MuFilter/UpstreamnSides"];}
        return conf_ints["MuFilter/DownstreamnSides"];
   }
+void MuFilter::getCoveredRuns(){
```
1. Is it absolutely necessary to have a method that only works via side-effects?
2. Since the method doesn't return anything, it should not be called `getCoveredRuns()`. Maybe something like `index_runs` or `parse_runs` would be better?
3. If you do insist on not using snake_case, I believe methods tend to use PascalCase in sndsw. Let's not also add camelCase to the mix.

cvilelahep replied:
1. I couldn't find a cleaner way of implementing this. I tried adding the code to the constructor or Initialize method and in both cases it runs before the conf_floats map gets populated and therefore it doesn't work. One alternative could be to move this side-effect to the InitEvent method, which has to be called anyway before the relevant methods get called. I'll see if this works.
2./3. If I can't resolve 1. I'll change to one of the suggestions in 2.

siilieva replied:
The conf_floats are populated when we do
`
import SndlhcGeo
geo = SndlhcGeo.GeoInterface(name_of_geoFile)
`

olantwin commented on shipLHC/MuFilter.h:
```diff
@@ -104,6 +104,12 @@ class MuFilter : public FairDetector
 		SNDLHCEventHeader *eventHeader;
 
+                // Vector to store runs covered in the geometry file.
+                std::vector<int> coveredRunsTimeAlignment;
```
Since it is mostly used for lookup, data structures other than vector might be better. Depends on the expected number of runs.

siilieva replied:
This object will have to host the re-alignment after each emulsion extraction i.e. it will have up to a few tens of members.
What do you envisage instead of a vector? It is rather fast to add elements and loop over vectors.

olantwin replied:
For a few tens of members, vectors are probably OK. For more, some hashtable probably would have been best?

olantwin commented on shipLHC/MuFilter.h:
```diff
@@ -104,6 +104,12 @@ class MuFilter : public FairDetector
 
+                // Vector to store runs covered in the geometry file.
+                std::vector<int> coveredRunsTimeAlignment;
+                TString lastTimeAlignmentTag;
```
We might want to consider using STL strings here. AFAIK they generally should be preferred nowadays. Using string_views could speed up some of the string operations.

siilieva replied:
(I think) we use the TString since allows easy sub-string access (and it can be interfaced with PyROOT with 0 pain)

olantwin replied:
string_view should be faster and allows very simple sub-string access.
I've never had issues with strings in pyroot (but plenty with TStrings!)

olantwin replied:
To elaborate, the tags should be strings, and string_views might speed up the string operations.

olantwin commented on shipLHC/MuFilter.h:
```diff
@@ -104,6 +104,12 @@ class MuFilter : public FairDetector
+                TString lastTimeAlignmentTag;
+                int lastRun;
+                bool alignmentInit;
+                void getCoveredRuns();
```
I'd prefer anything but camelCase here, unless you want to follow ROOT style fProperty naming.

olantwin reviewed (changes requested):
Comments apply to both, didn't make them a second time for SciFi.

Commits 2 and 4-5 could be fixups to 1 and 3 respectively.

Clearly seems like an improvement, but it could maybe still be optimised a bit more. I'll try to make some more concrete suggestions when I have time to look in more detail.

cvilelahep commented:
To respond to the question in #207 the way the improvement reported in this PR was calculated with the following script:
```python
import ROOT
import SndlhcGeo
import time
sndgeo = SndlhcGeo.GeoInterface('/eos/experiment/sndlhc/convertedData/physics/2023/geofile_sndlhc_TI18_V4_2023.root')

scifiDet = ROOT.gROOT.GetListOfGlobals().FindObject("Scifi")

ch = ROOT.TChain("rawConv")
ch.Add("/eos/experiment/sndlhc/convertedData/physics/2022/run_004541/sndsw_raw-0020.root")
ch.GetEntry(0)

N = 1000000
start = time.perf_counter()
for i in range(N) :
    scifiDet.InitEvent(ch.EventHeader)
    scifiDet.GetCorrectedTime(100000+20000+1000+115, 0, 0)
stop = time.perf_counter()

print("Seconds per call {}".format((stop - start)/N))
```

Which returns (best of 4 runs):
 - 7.34e-5 s per call before this PR
 - 6.91e-6 s per call with the modifications in this PR

olantwin commented:
> Which returns (best of 4 runs):
> 
>     * 7.34e-5 s per call before this PR
> 
>     * 6.91e-6 s per call with the modifications in this PR

Great, that's indeed quite a nice improvement!

cvilelahep commented:
I moved the run parsing to the InitEvent method and removed parse_runs.

I don't plan on further development at this point.

The comment on TString vs std::string above falls a bit outside the scope of this PR as the use of TString predates the PR and is widespread in sndsw.

siilieva commented:
This looks good to me. 
The first bullet in Issue #207 to create sub-maps once per det.InitEvent()  is addressed. The second one on map key (string, string_view TString, etc) remains.
@cvilelahep, could you squash the commits in a single one before we rebase?

cvilelahep commented:
Oh no... I think I messed this up. Let me try to fix it.

(I now officially hate anything that rewrites the commit history. Much prefer chaotic logs with thousands of inane commit messages :p )

cvilelahep commented:
OK, good to go now @siilieva

siilieva commented:
> OK, good to go now @siilieva

still @olantwin has to mark 'done'  changes he requested. I guess he's traveling now.

olantwin reviewed (approved):
Looks good enough for the time being.
